Tests/Orleans.Storage.Redis.Tests/StorageTests.cs

```diff
@@ -116,29 +115,62 @@ public StorageTests(ClusterFixture fixture)
+            var jsonSettings = new JsonSerializerSettings()
+            {
+                TypeNameHandling = TypeNameHandling.All,
+                PreserveReferencesHandling = PreserveReferencesHandling.Objects,
+                DateFormatHandling = DateFormatHandling.IsoDateFormat,
+                DefaultValueHandling = DefaultValueHandling.Ignore,
+                MissingMemberHandling = MissingMemberHandling.Ignore,
+                NullValueHandling = NullValueHandling.Ignore,
+                ConstructorHandling = ConstructorHandling.AllowNonPublicDefaultConstructor,
+            };
+            var now = DateTime.UtcNow;
+            var guid = Guid.NewGuid();
+            var state = new JsonTestGrainState
+            {
+                StringValue = "string value",
+                DateTimeValue = now,
+                GuidValue = guid,
+                IntValue = 12345,
+                GrainValue = _cluster.GrainFactory.GetGrain<IJsonTestGrain>(2222)
+            };
+            var testState = JsonConvert.SerializeObject(state, jsonSettings);
 
             var grain = _cluster.GrainFactory.GetGrain<IJsonTestGrain>(12345999);
             var grainRef = await grain.GetReference();
             var key = $"{grainRef.ToKeyString()}|json";
             await _fixture.Database.StringSetAsync(key, testState);
             
-            var guid = Guid.Parse("98d37743-457f-4ebd-8904-1eb366c6a95a");
             var result = await grain.Get();
-            Assert.Equal("string value", result.Item1);
-            Assert.Equal(12345, result.Item2);
-            Assert.NotEqual(default(DateTime), result.Item3);
-            Assert.Equal(guid, result.Item4);
-            Assert.Equal(2222, result.Item5.GetPrimaryKeyLong());
+            Assert.Equal(state.StringValue, result.Item1);
+            Assert.Equal(state.IntValue, result.Item2);
+            Assert.Equal(state.DateTimeValue, result.Item3);
+            Assert.Equal(state.GuidValue, result.Item4);
+            Assert.Equal(state.GrainValue.GetPrimaryKeyLong(), result.Item5.GetPrimaryKeyLong());
+        }
+
+        [Fact]
+        public async Task Json_Double_Activation_ETag_Conflict_Simulation()
+        {
+            var now = DateTime.UtcNow;
+            var guid = Guid.NewGuid();
+            var grain = _cluster.GrainFactory.GetGrain<IJsonTestGrain>(54321);
+            var grainRef = await grain.GetReference();
+
+            var stuff = await grain.Get();
+            var scheduler = TaskScheduler.Current;
 
-            await _fixture.Database.KeyDeleteAsync(key);
+            var key = $"{grainRef.ToKeyString()}|json";
+            await _fixture.Database.HashSetAsync(key, new[] { new HashEntry("etag", "derp") });
+            
+            var exception = await grain.Set("string value", 12345, now, guid, _cluster.GrainFactory.GetGrain<IJsonTestGrain>(2222));
```

Comment:
Added a test for ETag as well which I can't seem to get working. For some reason Orleans times out when calling `Set` here. When debugging I can see WriteStateAsync throwing the `ETagMismatchException` correctly, and `Set` returning it, but the call is never returned to the test and so Orleans times out after 30 seconds.

Reply:
> ` _cluster.GrainFactory.GetGrain<IJsonTestGrain>(2222)`

Shouldn't that has a `AsReference()` or something?